File: swift/Sources/DSA/Algorithms/InsertionSort.swift
/// Insertion sort algorithm.
/// Time: O(n²) worst/average, O(n) best (already sorted)
/// Space: O(1)
/// Stable: Yes
public enum InsertionSort {
    /// Sort an array in-place.
    public static func sort<T: Comparable>(_ array: inout [T], by comparator: ((T, T) -> Bool)? = nil) {
        let compare = comparator ?? { $0 < $1 }

        for i in 1..<array.count {
            let key = array[i]
            var j = i - 1

            while j >= 0 && compare(key, array[j]) {
                array[j + 1] = array[j]
                j -= 1
            }
            array[j + 1] = key
        }
    }

    /// Return a sorted copy of the array.
    public static func sorted<T: Comparable>(_ array: [T], by comparator: ((T, T) -> Bool)? = nil) -> [T] {
        var result = array
        sort(&result, by: comparator)
        return result
    }
}

File: java/src/main/java/com/dsa/algorithms/InsertionSort.java
package com.dsa.algorithms;

import java.util.Comparator;
import java.util.List;

/**
 * Insertion sort implementation.
 * Time: O(n^2) worst/average, O(n) best
 * Space: O(1)
 */
public class InsertionSort {

    @SuppressWarnings("unchecked")
    public static <T> void sort(List<T> arr) {
        sort(arr, (a, b) -> ((Comparable<T>) a).compareTo(b));
    }

    public static <T> void sort(List<T> arr, Comparator<T> comparator) {
        for (int i = 1; i < arr.size(); i++) {
            T current = arr.get(i);
            int j = i;

            while (j > 0 && comparator.compare(arr.get(j - 1), current) > 0) {
                arr.set(j, arr.get(j - 1));
                j--;
            }

            arr.set(j, current);
        }
    }

    public static <T> void sortDescending(List<T> arr, Comparator<T> comparator) {
        sort(arr, comparator.reversed());
    }
}
